src/group_theory/order_of_element.lean

```diff
@@ -373,6 +373,49 @@ end
 
 attribute [to_additive gsmul_eq_mod_add_order_of] gpow_eq_mod_order_of
 
+@[to_additive add_order_of_eq_zero_iff] lemma order_of_eq_zero_iff :
+  order_of x = 0 ↔ ¬ is_of_fin_order x :=
+⟨λ h H, (order_of_pos' H).ne' h, order_of_eq_zero⟩
+
+@[to_additive nsmul_inj_iff_of_add_order_of_eq_zero]
+lemma pow_inj_iff_of_order_of_eq_zero (h : order_of x = 0) {n m : ℕ} :
+  x ^ n = x ^ m ↔ n = m :=
+begin
+  by_cases hx : x = 1,
+  { rw [←order_of_eq_one_iff, h] at hx,
+    contradiction },
+  rw [order_of_eq_zero_iff, is_of_fin_order_iff_pow_eq_one] at h,
+  push_neg at h,
+  induction n with n IH generalizing m,
+  { cases m,
+    { simp },
+    { simpa [eq_comm] using h m.succ m.zero_lt_succ } },
+  { cases m,
+    { simpa using h n.succ n.zero_lt_succ },
+    { simp [pow_succ, IH] } }
+end
+
+lemma pow_inj_mod {n m : ℕ} :
+  x ^ n = x ^ m ↔ n % order_of x = m % order_of x :=
+begin
+  cases (order_of x).zero_le.eq_or_lt with hx hx,
+  { simp [pow_inj_iff_of_order_of_eq_zero, hx.symm] },
+  rw [pow_eq_mod_order_of, @pow_eq_mod_order_of _ _ _ m],
+  exact ⟨pow_injective_of_lt_order_of _ (nat.mod_lt _ hx) (nat.mod_lt _ hx), λ h, congr_arg _ h⟩
+end
+
+lemma nsmul_inj_mod {n m : ℕ} :
+  n • a = m • a ↔ n % add_order_of a = m % add_order_of a :=
+begin
+  cases (add_order_of a).zero_le.eq_or_lt with hx hx,
+  { simp [nsmul_inj_iff_of_add_order_of_eq_zero, hx.symm] },
+  rw [nsmul_eq_mod_add_order_of, @nsmul_eq_mod_add_order_of _ _ _ m],
+  refine ⟨nsmul_injective_of_lt_add_order_of a (nat.mod_lt n hx) (nat.mod_lt m hx), λ h, _⟩,
+  rw h
+end
+
+attribute [to_additive nsmul_inj_mod] pow_inj_mod
```

Comment:
Does this work directly on the lemmas now that `to_additive` knows about smul and pow? Or is that PR still unmerged?

Reply:
After merging master,
```
type mismatch at application
  @nsmul_injective_of_lt_add_order_of G x
    (n % @add_order_of G (@sub_neg_monoid.to_add_monoid G (@add_group.to_sub_neg_monoid G _inst_1)) x)
term
  n % @add_order_of G (@sub_neg_monoid.to_add_monoid G (@add_group.to_sub_neg_monoid G _inst_1)) x
has type
  ℕ
but is expected to have type
  add_left_cancel_monoid G
```
So it seems a separate lemma proof is necessary.

Reply:
@fpvandoorn, any ideas?